zhaquirks/philips/rdm002.py

```diff
@@ -0,0 +1,144 @@
+"""Signify RDM002 device."""
+from zigpy.profiles import zha
+from zigpy.quirks import CustomDevice
+from zigpy.zcl.clusters.general import (
+    Basic,
+    Groups,
+    Identify,
+    LevelControl,
+    OnOff,
+    Ota,
+    PowerConfiguration,
+    Scenes,
+)
+from zigpy.zcl.clusters.lightlink import LightLink
+
+from zhaquirks.const import (
+    BUTTON_1,
+    BUTTON_2,
+    BUTTON_3,
+    BUTTON_4,
+    COMMAND,
+    COMMAND_HOLD,
+    COMMAND_PRESS,
+    DEVICE_TYPE,
+    DOUBLE_PRESS,
+    ENDPOINTS,
+    INPUT_CLUSTERS,
+    LONG_RELEASE,
+    MODELS_INFO,
+    OUTPUT_CLUSTERS,
+    PROFILE_ID,
+    QUADRUPLE_PRESS,
+    QUINTUPLE_PRESS,
+    SHORT_PRESS,
+    SHORT_RELEASE,
+    TRIPLE_PRESS,
+)
+from zhaquirks.philips import (
+    PHILIPS,
+    SIGNIFY,
+    PhilipsBasicCluster,
+    PhilipsRemoteCluster,
+)
+
+
+class PhilipsRdmRemoteCluster(PhilipsRemoteCluster):
+    """Philips remote cluster for RDM devices."""
+
+    BUTTONS = {
+        1: BUTTON_1,
+        2: BUTTON_2,
+        3: BUTTON_3,
+        4: BUTTON_4,
+    }
+    PRESS_TYPES = {
+        0: COMMAND_PRESS,
+        1: COMMAND_HOLD,
+        2: SHORT_RELEASE,
+        3: LONG_RELEASE,
+    }
```

Comment:
The press types are exactly the same as in `PhilipsRemoteCluster`, right?
There, the constants aren't used though. Can you change that there (and remove the above code)?:
https://github.com/zigpy/zha-device-handlers/blob/ffdf2fbfe559f74e6fc00d61130b8025e18141ec/zhaquirks/philips/__init__.py#L162

Reply:
The `(SHORT|LONG)_RELEASE` constants used in `rdm002.py` are set to `"remote_button_short_release"` and `"remote_button_long_release"`, while the string literals `"short_release"` and `"long_release"` are used in `__init__.py`. `"short_release"` and `"long_release"` correspond to `COMMAND_M_SHORT_RELEASE` and `COMMAND_M_LONG_RELEASE`. Not entirely sure if something would break when changing the `PhilipsRemoteCluster`'s `PRESS_TYPE` values. Which constants would you prefer?

Also, I'm not sure if changing these values would break anything for existing users. Would it *not* break anything if we also change `PhilipsROM001`'s `device_automation_triggers` values?

Actually I was thinking about changing the implementation, introducing new `PhilipsRDMRemoteCluster` and `PhilipsRWLRemoteCluster` classes inheriting from `PhilipsRemoteCluster`, then change `PhilipsROM001` `PhilipsROM002` to inherit from `PhilipsRDMRemoteCluster`, and `PhilipsRWL022` from `PhilipsRWLRemoteCluster`, but as I'm not sure this can be done without breaking anything for existing users when changing things like button name and press type values, I haven't started working on this.

Servus MattWestb :)

Reply:
> The (SHORT|LONG)_RELEASE constants used in rdm002.py are set to "remote_button_short_release" and "remote_button_long_release", while the string literals "short_release" and "long_release" are used

Ah, I missed that.  The [`HUE_REMOTE_DEVICE_TRIGGERS`](https://github.com/zigpy/zha-device-handlers/blob/c01893fc3ed47bdce3f1d447ce13e18bb46ceec8/zhaquirks/philips/__init__.py#L38) that are used for all(?) other Hue remotes at the moment use those suffixes in the triggers.
IMO, the constants used in `rdm002.py` with `remote_button_xxx_release` aren't optimal. Personally, I think the press types that are currently used make more sense:
https://github.com/zigpy/zha-device-handlers/blob/ffdf2fbfe559f74e6fc00d61130b8025e18141ec/zhaquirks/philips/__init__.py#L162

Do you maybe want to remove the overriding `PRESS_TYPES` from your class, so we use all same press types for all devices?

And btw., your device triggers don't actually include short/long release ones for the buttons. Not sure if intended or not?

Reply:
> Actually I was thinking about changing the implementation, introducing new PhilipsRDMRemoteCluster...

Hmm, the current implementation isn't optimal, as `rdm001` also copies the complete custom cluster.
RWL022 also uses the "wrong button types".

If you want to have a go at changing the implementation, feel free to do so. If possible, we shouldn't really break device triggers / event data (used for blueprints).
The RWL022 event data is somewhat wrong though and should be changed maybe(?)

Reply:
Off the bat, sorry for the long post. I'd like to understand how I can help.

> Personally, I think the press types that are currently used make more sense: [...]

Is it OK to use constants here, or do you prefer string literals? Could still update `PRESS_TYPES` in `__init__.py` to use constants with the same value as the string literals currently used.

> And btw., your device triggers don't actually include short/long release ones for the buttons. Not sure if intended or not?

You're right. Also realized this and took a note below. Will have a look in the next couple of days.

> RWL022 also uses the "wrong button types".

What do you mean?

#### `HUE_REMOTE_DEVICE_TRIGGERS`
* ...is used by all RWL type devices, which are the 4-button dimmer remote controls.
* ROM001, the single button remote, is using separately defined `device_automation_triggers`, which is fine.
* And [RDM001](https://zigbee.blakadder.com/Philips_RDM001.html), the wall switch, is also using separately defined `device_automation_triggers`, but I think the trigger naming could be improved. I.e. instead of `TURN_ON`/`RIGHT`, to either use `LEFT`/`RIGHT` or `BUTTON_1`/`BUTTON_2`.

#### `PhilipsRemoteCluster`
* ... is defined in `__init__.py` and used directly for ROM and RWL devices.
* RDM002 currently inherits from it and overrides the `BUTTONS` and `PRESS_TYPES` without providing triggers for `LONG_PRESS` or `(SHORT|LONG)_RELEASE` events. (Need to check why i didn't add those - it's been some time since i wrote this).
* And finally we have RDM001, which defined it's own `PhilipsRemoteCluster`. Furthermore RDM001's `device_automation_triggers` lists multi-press triggers (e.g. double-, triple-, etc. -press events ) but I wonder how that works, as it's `PhilipsRemoteCluster` is missing the `send_press_event` function that generates similar events in `__init__.py`'s `PhilipsRemoteCluster`. For easy comparison of the two classes, I made [a diff here](https://gist.github.com/fholzer/f485009437d2cd50ef4265ca72e54490).

#### Questions
I would love to fix & streamline all that, but as i'm new to the code base i have a couple of questions:
* My understanding is that changing keys in the `device_automation_triggers` dict influences only how the triggers are presented in the UI, while changing the value (ie. `COMMAND`) would break things for existing users. Is that correct?
  * Even so, changing keys might break things that directly listen for zha_events and inspect the event arguments. Is that acceptable or not?
* Would you like to have a single `PhilipsRemoteCluster` defined in `__init__.py` and used all applicable devices instead of copying the class to device-specific files and use overrides where needed. Then all devices would support features like synthetic multi-press events.

Reply:
I still need to test all this, but already pushed my changes so you can have a look. The `device_automation_triggers` is unchanged for all existing devices.

Reply:
To get back to your original statement about `PRESS_TYPES` not being constants. Right, so in this new implementation I derive the `device_automation_triggers` from the `PRESS_TYPES`, and those `device_automation_triggers` were using these constants before.

Reply:
> Is it OK to use constants here, or do you prefer string literals? Could still update PRESS_TYPES in __init__.py to use constants with the same value as the string literals currently used.

No real preference for now. Constants are nicer if used in multiple places, but string literals should also be fine if it's used just once.

---

>> RWL022 also uses the "wrong button types".

> What do you mean?

The device has a shared on/off button (that's just an "on" button in the triggers + brightness buttons (fine) + a "hue" button ("off" button in the triggers).
It shouldn't be changed in this PR though, as it would break blueprints (react on `command` in `zha_event`) and device triggers, as those should also be "renamed".


--- 

> And [RDM001](https://zigbee.blakadder.com/Philips_RDM001.html), the wall switch, is also using separately defined device_automation_triggers, but I think the trigger naming could be improved. I.e. instead of TURN_ON/RIGHT, to either use LEFT/RIGHT or BUTTON_1/BUTTON_2.

Yes, it would break device triggers + blueprints though, as far as I can see.
All Hue remote related device triggers should probably be changed later in a separate PR (that's clearly marked as a breaking change). I'm not sure how much we're allowed to "break" now though, as there's some 6 month breaking change guideline in HA now. Not sure if/how that applies to libraries.

---

> And finally we have RDM001, which defined it's own PhilipsRemoteCluster. Furthermore RDM001's device_automation_triggers lists multi-press triggers (e.g. double-, triple-, etc. -press events ) but I wonder how that works, as it's PhilipsRemoteCluster is missing the send_press_event function that generates similar events in __init__.py's PhilipsRemoteCluster. For easy comparison of the two classes, I made [a diff here](https://gist.github.com/fholzer/f485009437d2cd50ef4265ca72e54490).

That's a good question on why those triggers are there if they can never be fired.

---

> My understanding is that changing keys in the device_automation_triggers dict influences only how the triggers are presented in the UI, while changing the value (ie. COMMAND) would break things for existing users. Is that correct?
Even so, changing keys might break things that directly listen for zha_events and inspect the event arguments. Is that acceptable or not?

I'd have to check at this point, but I think you're right. For device triggers to not break in automations, the keys (type + subtype) have to stay the same. Technically, the command can change IIRC.

A lot of blueprints listen to the `command` though and we shouldn't really break/change those for now, so (sadly) we should keep both the `command`and the device automation trigger the same. I'll need to properly look at your changes later, but do your current changes have both stay the same?

("Edit": Hmm, from a quick look, I think the commands are changed now, as both `BUTTONS` and `PRESS_TYPES` are changed.
Although I prefer your changes, commands (and device triggers) should stay the same for now (to be non-breaking).

(Also, for the "multi-press" stuff (ll `177-185` in `__init__.py`), should the appropriate press type come from `PRESS_TYPES` instead of string literals? Would be a bit cleaner maybe?)

---

> Would you like to have a single PhilipsRemoteCluster defined in __init__.py and used all applicable devices instead of copying the class to device-specific files and use overrides where needed. Then all devices would support features like synthetic multi-press events.

(I always found the idea of having synthetic multi-press events in the Philips quirks a bit weird, but I guess that's what we're doing for now.)

Since the class would mostly/somewhat be the same code for the devices, having one class in a central places makes the most sense to me. (Seems like that's what you did for now)

Reply:
> (Also, for the "multi-press" stuff (ll 177-185 in __init__.py), should the appropriate press type come from PRESS_TYPES instead of string literals? Would be a bit cleaner maybe?)

the PRESS_TYPES have keys that correspond to data we get in the event, so that dict can't be easily changes. The single/double/triple/...  entries are not subsequent entried, so the best we can do there is use the same constants  in the multipress part of the code that we use in the PRESS_TYPES like so: c2fdd4d